api/src/main/java/com/velocitypowered/api/event/connection/PreLoginEvent.java

```diff
@@ -46,4 +52,57 @@ public String toString() {
                 ", result=" + result +
                 '}';
     }
+    
+    /**
+     * Represents an "allowed/allowed with online mode/denied" result with a reason allowed for denial.
+     */
+    public static class PreLoginComponentResult extends ComponentResult {
+
+        private static final PreLoginComponentResult ALLOWED = new PreLoginComponentResult((Component) null);
+        private static final PreLoginComponentResult ALLOWED_ONLINEMODE = new PreLoginComponentResult(true);
+
+        private final boolean onlineMode;
+        /**
+         * Allows to enable a online mode for the player connection, when Velocity running in offline mode
+         * Does not have any sense if velocity running in onlineMode;
+         * @param onlineMode if true, offline uuid will be used for player connection if Velocity run in offlineMode
+         */
+        private PreLoginComponentResult(boolean onlineMode) {
+            super(true, null);
+            this.onlineMode = onlineMode;
+        }
+        
+        private PreLoginComponentResult(@Nullable Component reason) {
+            super(reason == null, reason);
+            // Don't care about this
+            this.onlineMode = false;
+        }
+
+        public boolean isOnlineMode() {
+            return this.onlineMode;
+        }
+
+        
+        @Override
+        public String toString() {
+            if (isOnlineMode()) {
+                return "allowed with online mode";
+            }
+            
+            return super.toString();
+        }
+
+        public static PreLoginComponentResult allowed() {
+            return ALLOWED;
+        }
+
+        public static PreLoginComponentResult allowedOnlineMode() {
```

Comment:
I would call this `forceOnlineMode()` and have an equivalent `forceOfflineMode()` function, too.

Reply:
I really can not understand why we need to have forceOfflineMode when proxy running in online mode.

Reply:
Okay, then we can omit it. If someone does need it we can add it.